README.md
# Bottom Drawer for Flutter [![Pub version](https://img.shields.io/pub/v/bottom_drawer.svg)](https://pub.dev/packages/bottom_drawer)

Bottom Drawer.

![DEMO](images/demo.gif)

### 1. Add dependency

```dart
dependencies:
  bottom_drawer: ^0.0.5
```

### 2. Import bottom drawer

```dart 
import 'package:bottom_drawer/bottom_drawer.dart';
```

### 3. Use bottom drawer

Create a bottom drawer controller.

```dart
/// create a bottom drawer controller to control the drawer.
BottomDrawerController controller = BottomDrawerController();
```

Build a bottom drawer widget.

```dart
/// return a bottom drawer widget.
Widget buildBottomDrawer(BuildContext context) {
  return BottomDrawer(
    /// your customized drawer header.
    header: Container(),
    /// your customized drawer body.
    body: Container(),
    /// your customized drawer header height.
    headerHeight: 60.0,
    /// your customized drawer body height.
    drawerHeight: 180.0,
    /// drawer background color.
    color: Colors.lightBlue,
    /// drawer controller.
    controller: controller,
  );
}
```

Control the bottom drawer.

```dart
/// open the bottom drawer.
controller.open();

/// close the bottom drawer.
controller.close();
```

For more information, see the [example](https://github.com/GP-Moon/bottom_drawer/tree/master/example).



CHANGELOG.md
## 0.0.5

* Fix bug.

## 0.0.4

* Add drawer BoxShadow.

## 0.0.3

* Null safety.

## 0.0.2

* Add some documentation comments.

## 0.0.1

* Bottom drawer widget.



test/bottom_drawer_test.dart
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('bottom drawer test', () {});
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:bottom_drawer_example/main.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('Verify demo started', (WidgetTester tester) async {
    await tester.pumpWidget(MyApp());
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text && (widget.data?.startsWith('Pressed') ?? false),
      ),
      findsOneWidget,
    );
  });
}


lib/bottom_drawer.dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

/// Created by GP
/// 2020/11/25.

/// drawer controller.
class BottomDrawerController {
  /// open drawer.
  void open() {
    _handler?.call(true);
  }

  /// close drawer.
  void close() {
    _handler?.call(false);
  }

  void Function(bool open)? _handler;
}

/// bottom drawer.
class BottomDrawer extends StatefulWidget {
  BottomDrawer({
    Key? key,
    required this.header,
    required this.body,
    required this.headerHeight,
    required this.drawerHeight,
    this.color = Colors.white,
    this.cornerRadius = 12,
    this.boxShadow = const [],
    this.duration = const Duration(milliseconds: 250),
    this.followTheBody = true,
    this.controller,
    this.callback,
  });

  @override
  _BottomDrawerState createState() => _BottomDrawerState();

  /// drawer header.
  final Widget header;

  /// drawer body.
  final Widget body;

  /// drawer header height.
  final double headerHeight;

  /// drawer total height.
  final double drawerHeight;

  /// drawer color.
  final Color color;

  /// drawer corner radius.
  final double cornerRadius;

  /// drawer animation duration.
  final Duration duration;

  /// drawer box shadow.
  final List<BoxShadow> boxShadow;

  /// If configured as true,
  /// When the drawer is open, the body slides to the top, and then slides down, the drawer will automatically close.
  /// When the drawer is closed, the body slides up and the drawer will automatically open.
  final bool followTheBody;

  /// drawer controller.
  final BottomDrawerController? controller;

  /// drawer status callback.
  final Function(bool opened)? callback;
}

class _BottomDrawerState extends State<BottomDrawer> with TickerProviderStateMixin {
  @override
  void initState() {
    super.initState();

    animationController = AnimationController(
      vsync: this,
      duration: widget.duration,
    );

    widget.controller?._handler = (open) {
      if (open)
        this.open(false);
      else
        this.close(false);
    };
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    if (mounted) {
      headerHeight = widget.headerHeight;
      drawerHeight = widget.drawerHeight;
      offset = drawerHeight - headerHeight;
    }
  }

  @override
  void dispose() {
    animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      width: double.infinity,
      height: double.infinity,
      alignment: Alignment.bottomCenter,
      child: buildSlider(),
    );
  }

  Widget buildSlider() {
    offset = offset.clamp(0.0, drawerHeight - headerHeight);
    return Transform.translate(
      offset: Offset(0.0, offset),
      child: RawGestureDetector(
        gestures: {
          VerticalDragGestureRecognizer: getRecognizer(),
        },
        child: buildDrawer(),
      ),
    );
  }

  GestureRecognizerFactoryWithHandlers<VerticalDragGestureRecognizer> getRecognizer() {
    return GestureRecognizerFactoryWithHandlers<VerticalDragGestureRecognizer>(
      _buildVerticalDragGestureRecognizerConstructor,
      _buildVerticalDragGestureRecognizerInitializer,
    );
  }

  VerticalDragGestureRecognizer _buildVerticalDragGestureRecognizerConstructor() {
    return VerticalDragGestureRecognizer();
  }

  void _buildVerticalDragGestureRecognizerInitializer(VerticalDragGestureRecognizer gestureRecognizer) {
    gestureRecognizer
      ..onStart = onDragStart
      ..onUpdate = onDragUpdate
      ..onEnd = onDragEnd;
  }

  void onDragStart(DragStartDetails details) {
    lastDrag = details.globalPosition.dy;
    lastDragOffset = 0;
    totalDragOffset = 0;
  }

  void onDragUpdate(DragUpdateDetails details) {
    lastDragOffset = details.globalPosition.dy - lastDrag;
    totalDragOffset += lastDragOffset;
    offset = offset + details.delta.dy;
    lastDrag = details.globalPosition.dy;
    setState(() {});
  }

  void onDragEnd(DragEndDetails details) {
    if (lastDragOffset < 0) {
      open();
    } else if (lastDragOffset > 0) {
      close();
    } else {
      if (totalDragOffset < 0) {
        open();
      } else if (totalDragOffset > 0) {
        close();
      }
    }
  }

  Widget buildDrawer() {
    return ConstrainedBox(
      constraints: BoxConstraints(
        maxHeight: drawerHeight,
        minHeight: headerHeight,
      ),
      child: Container(
        decoration: BoxDecoration(
          color: widget.color,
          borderRadius: BorderRadius.only(
            topLeft: Radius.circular(widget.cornerRadius),
            topRight: Radius.circular(widget.cornerRadius),
          ),
          boxShadow: widget.boxShadow,
        ),
        child: Column(
          children: [
            widget.header,
            Expanded(
              child: NotificationListener(
                onNotification: (Notification notification) => handleScrollNotification(notification),
                child: widget.body,
              ),
            ),
          ],
        ),
      ),
    );
  }

  bool handleScrollNotification(Notification notification) {
    if (widget.followTheBody)
      switch (notification.runtimeType) {
        case ScrollStartNotification:
          ScrollStartNotification scrollNotification = notification as ScrollStartNotification;
          ScrollMetrics metrics = scrollNotification.metrics;
          scrollOffset = metrics.pixels;
          scrollAtEdge = metrics.atEdge;
          return true;
        case ScrollUpdateNotification:
          ScrollUpdateNotification scrollNotification = notification as ScrollUpdateNotification;
          ScrollMetrics metrics = scrollNotification.metrics;
          double pixels = metrics.pixels;
          double flag = pixels - scrollOffset;
          if (flag > 0 && !opened) open();
          return true;
        case ScrollEndNotification:
          return true;
        case OverscrollNotification:
          OverscrollNotification scrollNotification = notification as OverscrollNotification;
          ScrollMetrics metrics = scrollNotification.metrics;
          double pixels = metrics.pixels;
          double flag = pixels - scrollOffset;
          if (scrollOffset == 0.0 && flag == 0.0 && scrollAtEdge && opened) close();
          return true;
      }
    return false;
  }

  void open([bool? callback]) {
    if (!opened) {
      double end = 0;
      double? start = offset;
      slide(start, end, callback ?? true);
    }
  }

  void close([bool? callback]) {
    if (opened) {
      double end = drawerHeight - headerHeight;
      double? start = offset;
      slide(start, end, callback ?? true);
    }
  }

  void slide(double? start, double end, bool callback) {
    opened = end == 0.0;

    if (callback) widget.callback?.call(opened);

    CurvedAnimation curve = new CurvedAnimation(
      parent: animationController,
      curve: Curves.easeOut,
    );

    animation = Tween(
      begin: start,
      end: end,
    ).animate(curve)
      ..addListener(() {
        setState(() {
          offset = animation.value;
        });
      });

    animationController.reset();
    animationController.forward();
  }

  bool opened = false;

  late double headerHeight;
  late double drawerHeight;

  double offset = 0.0;

  late double lastDrag;
  late double lastDragOffset;
  late double totalDragOffset;

  double scrollOffset = 0.0;
  bool scrollAtEdge = false;

  late Animation<double> animation;
  late AnimationController animationController;
}


example/lib/main.dart
import 'package:bottom_drawer/bottom_drawer.dart';
import 'package:flutter/material.dart';

/// Created by GP
/// 2020/11/25.

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Bottom drawer example app'),
        ),
        body: Stack(
          children: [
            Column(
              mainAxisAlignment: MainAxisAlignment.center,
              crossAxisAlignment: CrossAxisAlignment.center,
              children: [
                Text('Pressed $_button'),
                Padding(
                  padding: EdgeInsets.only(top: 30.0),
                  child: Row(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      IconButton(
                        icon: Icon(
                          Icons.arrow_upward,
                          color: Colors.black,
                        ),
                        onPressed: () {
                          _controller.open();
                          setState(() {
                            _button = 'Open Drawer';
                          });
                        },
                      ),
                      Divider(
                        height: 10.0,
                      ),
                      IconButton(
                        icon: Icon(
                          Icons.arrow_downward,
                          color: Colors.black,
                        ),
                        onPressed: () {
                          _controller.close();
                          setState(() {
                            _button = 'Close Drawer';
                          });
                        },
                      ),
                    ],
                  ),
                ),
              ],
            ),
            _buildBottomDrawer(context),
          ],
        ),
      ),
    );
  }

  Widget _buildBottomDrawer(BuildContext context) {
    return BottomDrawer(
      header: _buildBottomDrawerHead(context),
      body: _buildBottomDrawerBody(context),
      headerHeight: _headerHeight,
      drawerHeight: _bodyHeight,
      color: Colors.lightBlue,
      controller: _controller,
      boxShadow: [
        BoxShadow(
          color: Colors.black.withOpacity(0.15),
          blurRadius: 60,
          spreadRadius: 5,
          offset: const Offset(2, -6), // changes position of shadow
        ),
      ],
    );
  }

  Widget _buildBottomDrawerHead(BuildContext context) {
    return Container(
      height: _headerHeight,
      child: Column(
        children: [
          Padding(
            padding: EdgeInsets.only(
              left: 10.0,
              right: 10.0,
              top: 10.0,
            ),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: _buildButtons('', 1, 2),
            ),
          ),
          Spacer(),
          Divider(
            height: 1.0,
            color: Colors.grey,
          ),
        ],
      ),
    );
  }

  Widget _buildBottomDrawerBody(BuildContext context) {
    return Container(
      width: double.infinity,
      height: _bodyHeight,
      child: SingleChildScrollView(
        child: Column(
          children: _buildButtons('Body', 1, 25),
        ),
      ),
    );
  }

  List<Widget> _buildButtons(String prefix, int start, int end) {
    List<Widget> buttons = [];
    for (int i = start; i <= end; i++)
      buttons.add(TextButton(
        child: Text(
          '$prefix Button $i',
          style: TextStyle(
            fontSize: 15.0,
            color: Colors.black,
          ),
        ),
        onPressed: () {
          setState(() {
            _button = '$prefix Button $i';
          });
        },
      ));
    return buttons;
  }

  String _button = 'None';
  double _headerHeight = 60.0;
  double _bodyHeight = 180.0;
  BottomDrawerController _controller = BottomDrawerController();
}
